### PR TITLE
TICKF calculatePoolDistr

### LINKED ISSUE #3132: earlier thunk for the per-stake-pool stake distribution
This is a stop-gap measure for the problem described in #3034.

Create the thunk for the `calculatePoolDistr` when the corresponding snapshot is created (when crossing the previous epoch boundary), not in the `NEWEPOCH` rule. This way, when the consensus layer forecasts across the boundary multiple times, the expensive computation only happens once per epoch boundary per node per fork.

### PR DESCRIPTION
**Pull request by aniketd**
- [ ] Get approval about exhaustiveness of changes required
- [x] Revert `pattern` implementation and move `calc` from within `SnapShotRaw` to within `SnapShots` as it was earlier.
- [ ] Update changelog
- [ ] Squash all history and write an elaborate commit message with links to issues

References
- https://github.com/input-output-hk/cardano-node/issues/4421
- #3034 
- #3132 

Closes #3132

### CODE REVIEW

**Comment by aniketd** [comment]
@nfrisby I have just merged your latest changes on to this one, which remains rebased on top of `master`. Would really like to have your eyes go over this.

File: cabal.project [hunk 1]
```diff
@@ -62,6 +62,72 @@ benchmarks: true
 -- The only sensible test display option
 test-show-details: streaming
 
+-- <<<<<<< HEAD
```
Comment (lehins) [inline cabal.project]: Conflict

File: eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs [hunk 2]
```diff
@@ -153,7 +153,7 @@ instance Crypto c => TranslateEra (MaryEra c) ShelleyTxOut where
 
 instance Crypto c => TranslateEra (MaryEra c) UTxO where
   translateEra ctxt utxo =
-    return $ UTxO (translateEra' ctxt <$> unUTxO utxo)
+    return $ UTxO (translateEra' ctxt `Map.map` unUTxO utxo)
```
Comment (lehins) [inline eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: `Map` has `Functor` instance, why this change?

File: cabal.project (hunk 1 above)
Comment (aniketd) [reply cabal.project]: Yes :slightly_smiling_face: , I kept it just in case they were needed from the original PR. Going to remove them in case they aren't needed. Tests seem to be failing. Will look closer what's happening.

File: eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs (hunk 2 above)
Comment (aniketd) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: This change was part of the original PR. Will check the implementation and then revert this change if it is unnecessary :+1:

Comment (aniketd) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: `fmap` is indeed defined in terms of `map`, for both `v0.6.6` as well as `0.6.5.1` of `containers`. Reverting this change. :+1:

Comment (amesgen) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: @nfrisby might clarify, but I think he did this to prevent extra thunks as the `Functor Map` instance is lazy in the values, while `Data.Map.Strict.map` is strict.

Comment (lehins) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: @amesgen That is very good point. We probably want to add a comment about this, so no-one switches it back to `fmap` in the future.

Comment (aniketd) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: @amesgen @nfrisby sorry didn't catch that while I was looking. Thanks for the clarification!

Comment (lehins) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: Here is a separate PR: https://github.com/input-output-hk/cardano-ledger/pull/3143

Comment (nfrisby) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Translation.hs]: Yes, that diff should only be in PR 3143, not this one. It was merely a "leftover" in my commits.

**Review by aniketd — COMMENTED** [review]
@nfrisby this commit moves the `PoolDistr` from the record `SnapShots` to inside each individual `SnapShot` as suggested by @TimSheard. I'd be glad to receive both your reviews on this change.

**Comment by aniketd** [comment]
@JaredCorduan I have updated this PR to reflect our latest discussion.
1. Move `PoolDistr` back to `SnapShots` from `SnapShot`
2. Remove `SnapShotRaw` and corresponding `pattern synonym`

This is ready for your review!

**Comment by JaredCorduan** [comment]
replaced by #3209
